crates/core/src/network/server_builder.rs

```diff
@@ -71,37 +68,37 @@ impl NetworkServerBuilder {
         bind_address: &BindAddress,
     ) -> Result<(), anyhow::Error> {
         node_rpc_health.update(NodeRpcStatus::StartingUp);
-        // Trace layer
-        let span_factory = tower_http::trace::DefaultMakeSpan::new()
+
+        // Trace layer for HTTP requests
+        let http_span_factory = tower_http::trace::DefaultMakeSpan::new()
             .include_headers(true)
             .level(tracing::Level::ERROR);
 
-        let axum_router = self
-            .axum_router
-            .unwrap_or_default()
-            .layer(TraceLayer::new_for_http().make_span_with(span_factory.clone()))
-            .fallback(handler_404);
+        // Prepare Axum router for HTTP requests
+        let axum_router = self.axum_router.unwrap_or_default();
+
+        // Prepare gRPC routes
+        let mut grpc_routes = self.grpc_routes.unwrap_or_default();
 
+        // Add reflection service to gRPC routes
         let mut reflection_service_builder = tonic_reflection::server::Builder::configure();
         for descriptor in self.grpc_descriptors {
             reflection_service_builder =
                 reflection_service_builder.register_encoded_file_descriptor_set(descriptor);
         }
+        grpc_routes = grpc_routes.add_service(reflection_service_builder.build_v1()?);
+
+        // Convert gRPC routes to Axum router
+        let grpc_router = grpc_routes.prepare().into_axum_router();
+
+        // Merge HTTP and gRPC routers
+        let combined_router = axum_router
+            .merge(grpc_router)
+            .fallback(handler_404)
+            .layer(TraceLayer::new_for_http().make_span_with(http_span_factory));
```

Comment:
Does that mean that we don't want `TraceLayer::new_for_grpc()` for grpc routes?

Reply:
Good catch, I have missed this one out.